internal/fs/vss_windows.go

```diff
@@ -0,0 +1,986 @@
+	addToSnapshotSet              uintptr
+	doSnapshotSet                 uintptr
+	deleteSnapshots               uintptr
+	importSnapshots               uintptr
+	breakSnapshotSet              uintptr
+	getSnapshotProperties         uintptr
+	query                         uintptr
+	isVolumeSupported             uintptr
+	disableWriterClasses          uintptr
+	enableWriterClasses           uintptr
+	disableWriterInstances        uintptr
+	exposeSnapshot                uintptr
+	revertToSnapshot              uintptr
+	queryRevertStatus             uintptr
+}
+
+// getVTable returns the vtable for IVssBackupComponents.
+func (vss *IVssBackupComponents) getVTable() *IVssBackupComponentsVTable {
+	return (*IVssBackupComponentsVTable)(unsafe.Pointer(vss.RawVTable))
+}
+
+// AbortBackup calls the equivalent VSS api.
+func (vss *IVssBackupComponents) AbortBackup() error {
+	result, _, _ := syscall.Syscall(vss.getVTable().abortBackup, 1,
+		uintptr(unsafe.Pointer(vss)), 0, 0)
+
+	return newVssErrorIfResultNotOK("AbortBackup() failed", HRESULT(result))
+}
+
+// InitializeForBackup calls the equivalent VSS api.
+func (vss *IVssBackupComponents) InitializeForBackup() error {
+	result, _, _ := syscall.Syscall(vss.getVTable().initializeForBackup, 2,
+		uintptr(unsafe.Pointer(vss)), 0, 0)
+
+	return newVssErrorIfResultNotOK("InitializeForBackup() failed", HRESULT(result))
+}
+
+// SetContext calls the equivalent VSS api.
+func (vss *IVssBackupComponents) SetContext(context VssContext) error {
+	result, _, _ := syscall.Syscall(vss.getVTable().setContext, 2,
+		uintptr(unsafe.Pointer(vss)), uintptr(context), 0)
+
+	return newVssErrorIfResultNotOK("SetContext() failed", HRESULT(result))
+}
+
+// GatherWriterMetadata calls the equivalent VSS api.
+func (vss *IVssBackupComponents) GatherWriterMetadata() (*IVSSAsync, error) {
+	var oleIUnknown *ole.IUnknown
+	result, _, _ := syscall.Syscall(vss.getVTable().gatherWriterMetadata, 2,
+		uintptr(unsafe.Pointer(vss)), uintptr(unsafe.Pointer(&oleIUnknown)), 0)
+
+	err := newVssErrorIfResultNotOK("GatherWriterMetadata() failed", HRESULT(result))
+	return vss.convertToVSSAsync(oleIUnknown, err)
+}
+
+// convertToVSSAsync looks up IVSSAsync interface if given result
+// is a success.
+func (vss *IVssBackupComponents) convertToVSSAsync(
+	oleIUnknown *ole.IUnknown, err error) (*IVSSAsync, error) {
+	if err != nil {
+		return nil, err
+	} else {
+		comInterface, err := queryInterface(oleIUnknown, UIID_IVSS_ASYNC)
+		if err != nil {
+			return nil, err
+		}
+
+		iVssAsync := (*IVSSAsync)(unsafe.Pointer(comInterface))
+		return iVssAsync, nil
+	}
+}
+
+// IsVolumeSupported calls the equivalent VSS api.
+func (vss *IVssBackupComponents) IsVolumeSupported(volumeName string) (bool, error) {
+	volumeNamePointer, err := syscall.UTF16PtrFromString(volumeName)
+	if err != nil {
+		panic(err)
+	}
+
+	isSupported := false
+	var result uintptr = 0
+
+	if runtime.GOARCH == "386" {
+		ptrToID := uintptr(unsafe.Pointer(ole.IID_NULL))
```

Comment:
According to this [comments](https://github.com/golang/go/blob/9499a2e10840e3878bff06e7f9a3bdaa56c5ca45/src/unsafe/unsafe.go#L94) `uintptr` cannot be stored in variable. I think this (and other three places) should be:
```
                id := (*[4]uintptr)(unsafe.Pointer(ole.IID_NULL))
                result, _, _ = syscall.Syscall9(vss.getVTable().isVolumeSupported, 7,
                        uintptr(unsafe.Pointer(vss)), id[0], id[1], id[2], id[3],
                        uintptr(unsafe.Pointer(volumeNamePointer)), uintptr(unsafe.Pointer(&isSupported)), 0, 0)
```
Additionally it's more readable.

Reply:
If I'm not completely mistaken, the rules for uintptr and unsafe code are necessary to ensure that the garbage collection work properly. The `uintptr(unsafe.Pointer(ole.IID_NULL))` would break when `ole.IID_NULL` (or whatever variable is used at that point) wouldn't be referenced for anywhere else anymore. After as I remember the code doesn't have that garbage collection problem, but I agree that your proposed change is far more readable.

Reply:
>After as I remember the code doesn't have that garbage collection problem

In this case you right: `ole.IID_NULL` is global variable and newer be garbage collected.
But in other similar cases code translate address of local var to `uintptr` and after that this variable never references by it's name only by conversion of this `uintptr` to `unsafe.Pointer` with subsequent dereferencing. Smart compiler may reuse memory address of this variable immediately after conversion to `uintptr`. I don't think GO compiler do it now, but:
>Code not using these patterns is likely to be invalid today or to become invalid in the future.

Reply:
As the `snapshotID` variable in the other two cases is passed on the stack, it shouldn't be affected by garbage collection (or does the go compiler move that variable to the heap?). But anyways I agree that this should be cleaned up.

Reply:
>As the snapshotID variable in the other two cases is passed on the stack, it shouldn't be affected by garbage collection

In this cases I mean [stack reuse optimization](https://gcc.gnu.org/onlinedocs/gcc/Code-Gen-Options.html#index-fstack_005freuse), not garbage collection. And as I said, this optimization is not present in the current version of Go compiler.

Reply:
`go vet` also don't like this constructions:
```
.\vss_windows.go:298:54: possible misuse of unsafe.Pointer
.\vss_windows.go:299:24: possible misuse of unsafe.Pointer
.\vss_windows.go:300:24: possible misuse of unsafe.Pointer
.\vss_windows.go:301:24: possible misuse of unsafe.Pointer
.\vss_windows.go:336:24: possible misuse of unsafe.Pointer
....
```

Reply:
Fixed in https://github.com/restic/restic/pull/2274/commits/97322db98cdc31a9aba2c358e59809418cdf68be